Prevent UOE when MessageBodyReader can handle multiple media types

geoand opened the pull request:
The basic idea of the PR is to get rid of the useless usages of `MediaTypeHelper#getBestMatch`.
Because of that we can now enforce the immutability of the media types in
`ResourceReader` and `ResourceWriter`. This change is also beneficial from a performance standpoint.
The places that do indeed need `getBestMatch` are very unlikely to be on the hot path, so I went ahead and made that method use a copy of the list and operate on it.

Furthermore the PR applies some extra polish to MediaTypeHelper and
does away with the Iterator allocation in during the lookup
of the appropriate `MessageBodyReader` and `MessageBodyWriter classes`

Related to: https://github.com/quarkusio/quarkus/pull/13997#issuecomment-767260680

FroMage commented on independent-projects/resteasy-reactive/common/runtime/src/main/java/org/jboss/resteasy/reactive/common/core/Serialisers.java:
```diff
@@ -91,11 +91,12 @@ private void readerLookup(MediaType mediaType, RuntimeType runtimeType, List<Med
                     continue;
                 }
-                MediaType match = MediaTypeHelper.getBestMatch(mt, goodTypeReader.mediaTypes());
+                MediaType match = MediaTypeHelper.getBestMatch(mt, goodTypeReader.modifiableMediaTypes());
```
So `MediaTypeHelper.getBestMatch` modifies its parameter? And `modifiableMediaTypes` provides a back-door to its immutable view? I can't recall the lifecycle of `ResourceReader/Writer`, but if it's application-long, it doesn't seem like a good idea to modify it while looking up reader/writers, no?

Probably I didn't understand something :)

geoand replied:
Yes, `bestMatch` does modify the parameters. But `modifiableMediaTypes` doesn't allow you to change the original order in any way

gsmet commented:
It might be a stupid idea but why don't you sort them once and for all before making the list immutable?

Or do you expect the list to be sorted in different ways depending on the context?

geoand commented:
> It might be a stupid idea but why don't you sort them once and for all before making the list immutable?
> 

Although this makes sense, it won't work because the `ResourceReaderComparator` needs to compare the _original_ media types (basically it does this because these are taken as the declaration of how imporant the `MessageBodyReader` to that media type - if that explanation makes any sense)

> Or do you expect the list to be sorted in different ways depending on the context?

geoand commented:
But this indeed warrants another look into how the readers are determined (although I really hate this part as there are so many TCK tests that fail even when the slightest change is made...)

geoand committed:
Prevent UOE when MessageBodyReader can handle multiple media types

The basic idea of the PR is to get rid of the useless usages of MediaTypeHelper.getBestMatch.
Because of that we can now enforce the immutability of the media types in
ResourceReader and ResourceWriter.

Furthermore the PR applies some extra polish to MediaTypeHelper and
does away with the Iterator allocation in during the lookup
of the appropriate MessageBodyReader and MessageBodyWriter classes

Related to: https://github.com/quarkusio/quarkus/pull/13997#issuecomment-767260680

geoand commented:
I totally changed the PR and also updated the description.

gsmet commented on independent-projects/resteasy-reactive/common/runtime/src/main/java/org/jboss/resteasy/reactive/common/util/MediaTypeHelper.java:
```diff
@@ -133,19 +126,31 @@ public static int compareWeight(MediaType one, MediaType two) {
+        if (!hasAtMostOneItem(provided)) {
+            provided = new ArrayList<>(provided);
+            sortByWeight(provided);
+        }
```
This is never used in a hot path? Because I could see how the extra allocations could slow things down.

geoand replied:
It is not, because I removed the calls to `getBestMatch` from the hot path. The remaining calls to it are obscure features
